File: lib/src/main/java/org/simplepresenter/Presenter.java
package org.simplepresenter;

import java.lang.ref.WeakReference;
import java.util.LinkedList;
import java.util.List;

public class Presenter implements ObservablePresenter {

    private WeakReference<ViewDelegate<?>> delegateReference;

    private List<ViewCommand> commands = new LinkedList<>();
    private ViewCommand lastCommand;

    private CommandObserver observer;

    private CommandObserver getObserver() {
        if (observer == null) {
            observer = createCommandObserver();
        }
        return observer;
    }

    public CommandObserver createCommandObserver() {
        return new CommandObserver(this);
    }

    public void onViewCreated() {
    }

    public void onViewDestroyed() {
    }

    public final void applyViewState(ViewCommand command) {
        getObserver().beforeApplyCommand(command);
        commands.add(command);
        lastCommand = command;
        if (isViewResumed()) {
            dispatchCommand(command);
        }
    }

    private boolean isViewResumed() {
        return delegateReference != null && delegateReference.get() != null && delegateReference.get().isResumed();
    }

    public void bind(ViewDelegate<?> viewDelegate) {
        clearReferenceToView();
        delegateReference = new WeakReference<ViewDelegate<?>>(viewDelegate);
    }

    void onRestoreView() {
        for (int i = 0; i < commands.size(); i++) {
            dispatchCommand(commands.get(i));
        }
    }

    public void dispatchCommand(ViewCommand command) {
        if (delegateReference != null && delegateReference.get() != null) {
            delegateReference.get().delegateCommand(command);
            getObserver().afterCommandDispatched(command);
        }
    }

    public void clearReferenceToView() {
        if (delegateReference != null) {
            delegateReference.clear();
        }
        delegateReference = null;
    }

    List<ViewCommand> getCommands() {
        return commands;
    }

    @Override
    public final void applyCommands(Action f) {
        commands = f.invoke(commands);
    }
}

File: lib/src/main/java/org/simplepresenter/Presenter.kt
package org.simplepresenter

import org.simplepresenter.behavior.DelegatedBehavior
import java.lang.ref.WeakReference
import java.util.*

open class Presenter {

    private val behavior: DelegatedBehavior
    private var viewRef: WeakReference<ViewDelegate>? = null

    var commands: List<ViewCommand> = LinkedList()

    init {
        behavior = DelegatedBehavior(CommandEngine.Bridge.currentEngine.behaviorFactory.behaviors)
    }

    // public api -->

    open fun onViewCreated() {
    }

    open fun onViewDestroyed() {
    }

    fun applyCommand(command: ViewCommand) {
        commands = behavior.beforeApply(command, commands) + command
        if (isViewResumed) {
            dispatchCommand(command)
        }
    }

    // internal api -->

    val isViewResumed: Boolean
        get() = viewRef?.get()?.isResumed ?: false

    fun bind(viewDelegate: ViewDelegate) {
        clearReferenceToView()
        viewRef = WeakReference(viewDelegate)
    }

    fun onRestoreView() = commands.forEach { dispatchCommand(it) }

    fun dispatchCommand(command: ViewCommand) {
        viewRef?.get()?.delegateCommand(command)
        commands = behavior.afterDispatched(command, commands)
    }

    fun clearReferenceToView() {
        viewRef?.clear()
        viewRef = null
    }
}